projects/laji/src/app/shared-modules/audio-viewer/audio-viewer/audio-spectrogram/audio-spectrogram.component.ts

```diff
@@ -0,0 +1,105 @@
+import {
+  Component,
+  Input,
+  AfterViewInit,
+  OnChanges,
+  ElementRef,
+  ViewChild,
+  HostListener,
+  Output,
+  EventEmitter,
+  ChangeDetectionStrategy,
+  ChangeDetectorRef
+} from '@angular/core';
+
+import { AudioViewerMode, IAudioViewerArea, IAudioViewerRectangle, ISpectrogramConfig } from '../../models';
+import { timer } from 'rxjs';
+import { take } from 'rxjs/operators';
+
+@Component({
+  selector: 'laji-audio-spectrogram',
+  templateUrl: './audio-spectrogram.component.html',
+  styleUrls: ['./audio-spectrogram.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class AudioSpectrogramComponent implements AfterViewInit, OnChanges {
+  @ViewChild('container', {static: true}) containerRef: ElementRef<HTMLDivElement>;
+
+  @Input() buffer: AudioBuffer;
+  @Input() view: IAudioViewerArea;
+  @Input() defaultView: IAudioViewerArea;
+
+  @Input() focusArea: IAudioViewerArea;
+  @Input() highlightFocusArea = false;
+  @Input() onlyFocusAreaClickable = false;
+  @Input() showAxisLabels = true;
+  @Input() axisFontSize = 10;
+  @Input() rectangles: IAudioViewerRectangle[];
+
+  @Input() pregeneratedSpectrogramUrl?: string;
+  @Input() config: ISpectrogramConfig;
+
+  @Input() currentTime: number;
+  @Input() mode: AudioViewerMode;
+
+  @Input() width: number;
+  @Input() height: number;
+  @Input() margin: { top: number, bottom: number, left: number, right: number };
+
+  @Output() spectrogramReady = new EventEmitter();
+  @Output() dragStart = new EventEmitter();
+  @Output() dragEnd = new EventEmitter<number>();
+  @Output() spectrogramClick = new EventEmitter<number>();
+  @Output() spectrogramDblclick = new EventEmitter<number>();
+  @Output() zoomEnd = new EventEmitter<IAudioViewerArea>();
+  @Output() drawEnd = new EventEmitter<IAudioViewerArea>();
+
+  _width: number;
+  _height: number;
+  _margin: { top: number, bottom: number, left: number, right: number };
+
+  private defaultMargin = { top: 10, bottom: 40, left: 50, right: 10 };
+  private defaultMarginWithoutLabels = { top: 10, bottom: 20, left: 30, right: 10 };
+
+  constructor(
+    private cdr: ChangeDetectorRef
+  ) {
+    this.updateMargin();
+  }
+
+  @HostListener('window:resize')
+  onResize() {
+    this.updateWidthAndHeigth();
+  }
+
+  ngAfterViewInit() {
+    // update width and height immediately and after some time has passed to ensure that the content has loaded
+    timer(0, 300).pipe(take(2)).subscribe(() => {
+      this.onResize();
+      this.cdr.markForCheck();
+    });
```

Comment:
Is the timer really necessary? I tried removing it and the spectogram view initialized just fine. Is there some race condition that occurs on slower machines?

Reply:
It's necessary, `this.containerRef.nativeElement.offsetWidth` is not always correct at `ngAfterViewInit`. It happens more frequently when you change between views (double click some template/candidate and then press cancel).

Reply:
How about updating on `laji-spectogram` on `spectogramReady`? I tested at seems to work. Btw, the `ngOnChanges` stuff should maybe be done on `ngAfterViewChecked`.

Reply:
It may work because `laji-spectrogram` has a delay before it emits `spectrogramReady` (line 60: `observable.pipe(delay(0))`). It has that because sometimes the spectrogram is ready straight away and you couldn't emit an output on the `ngOnChanges` function (not sure if it's fixed in the current Angular version). I wouldn't add it to that because it may be confusing that it depends on another component.

> Btw, the ngOnChanges stuff should maybe be done on ngAfterViewChecked.

I think `ngOnChanges` is ok place because it reacts to changes in `Input` variables. I did small refactoring for the component [here](https://github.com/luomus/laji/pull/219/commits/a6baa4331b8cce67672169e4f6f7ec0f83d1185a) to make it hopefully little clearer

Reply:
Hmm. I didn't notice the `delay(0)`. I tried to implement this without the delay and the `timer`, and rely on hooks (`AfterViewChecked`) but didn't get it to work. The canvas & img drawing should be synchronous browser stuff, so not sure what's going on there. Well, I've spent enough time on this now - if it works it works... :D if you know why the `delay(0)` is needed, could you please make a comment for that?

Reply:
I added a comment but I'm not sure what is the reason for that. I remember it worked ok without SSR before and only SSR was causing problems. Now when I tested it doesn't work even without SSR if there is no delay...